File: mpcc_trajectory/examples/spline_trajectory_generator.py
#!/usr/bin/env python3

# Copyright 2025 Universidad Politécnica de Madrid
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
#    * Redistributions of source code must retain the above copyright
#      notice, this list of conditions and the following disclaimer.
#
#    * Redistributions in binary form must reproduce the above copyright
#      notice, this list of conditions and the following disclaimer in the
#      documentation and/or other materials provided with the distribution.
#
#    * Neither the name of the Universidad Politécnica de Madrid nor the names of its
#      contributors may be used to endorse or promote products derived from
#      this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
"""Spline trajectory generator."""

__authors__ = 'Carmen De Rojas Pita-Romero'
__copyright__ = 'Copyright (c) 2025 Universidad Politécnica de Madrid'
__license__ = 'BSD-3-Clause'

from hermite_spline import HermiteSpline, compute_arc_length_reparametrization, get_segment_from_arc_length
from typing import List, Tuple, Dict
import numpy as np
import matplotlib.pyplot as plt

class SplineTrajectoryGenerator:
    """Spline trajectory generator base class."""
        
    def __init__(
        self,
        wp: dict,
        origin: dict,
        num_wp: int,
        cyclic: bool,
    ):
        """
        :param wp: dict of waypoints and tangents
        :param origin: dict position and tangent of the drone"""
        # Global parameters

        self.drone_pose = origin 
        self.num_wp = num_wp
        self.cyclic =cyclic
        self.theta = None
        self.spline_reparametrization = None
        self.theta_intervals = None

        # Add key to waypoints dict
        self.waypoints = {
            f'wp_{i}': {'waypoints': pos, 'waypoints_tg': tg}
            for i, (pos, tg) in enumerate(zip(wp['waypoints'], wp['waypoints_tg']))
        }
        
    def first_spline(self):

        """Generate the first spline trajectory."""

        wps = [self.drone_pose['waypoints']] + [self.waypoints[f'wp_{i}']['waypoints'] for i in range(self.num_wp-1 )]
        tgs = [self.drone_pose['waypoints_tg']] + [self.waypoints[f'wp_{i}']['waypoints_tg'] for i in range(self.num_wp -1)]
        self.generate_spline(wps, tgs)

    def generate_spline(self, point: list[np.ndarray],tg: list[np.ndarray]):

        """Generate spline trajectory from waypoints and tangents."""

        # print(f"Generating spline with points: {point} and tangents: {tg}\n")
        spline = HermiteSpline(point, tg)
        self.spline_reparametrization = compute_arc_length_reparametrization(spline, n_samples=200, poly_degree=5)
        print(f"Spline generated with total length: {self.spline_reparametrization['total_length']}\n")
        print(f"Spline generated with total length: {self.spline_reparametrization['ti']}\n")
    
    def check_status_trajectory(self, s_eval,)-> bool:

        """Check the status of the trajectory and regenerate if needed."""

        segment_idx, t_val, s_norm = get_segment_from_arc_length(s_eval, self.spline_reparametrization['ti'],
            self.spline_reparametrization['total_length'],
            self.spline_reparametrization['poly_coeffs']
            )
        if segment_idx !=0:
            print(f"Segment idx: {segment_idx}, t_val: {t_val}, s_norm: {s_norm}\n")
            print('Regenerating spline...\n')
            wps = [data['waypoints'] for data in list(self.waypoints.values())[:self.num_wp ]]
            tgs = [data['waypoints_tg'] for data in list(self.waypoints.values())[:self.num_wp ]]
            if not self.cyclic:
                if len(self.waypoints) < self.num_wp:
                    return False
            self.generate_spline(wps, tgs)

            # Remove the first waypoint and add it to the end
            first_wp_key, first_wp_value = list(self.waypoints.items())[0]
            self.waypoints.pop(first_wp_key)

            if  self.cyclic:
                self.waypoints[first_wp_key] = first_wp_value
                return True
            return True
        return False

def main():
    drone_pose = {
        'waypoints': np.array([0.0, 0.0, 0.0]),
        'waypoints_tg': np.array([1.0, 0.0, 0.0])  
    }

    waypoints = {
        'waypoints': [
            [0.0, 0.0, 0.5],
            [1.0, 0.0, 1.0],
            [2.0, 0.0, 1.0],
            [3.0, 0.0, 1.0],
            [4.0, 0.0, 1.0],
            [5.0, 0.0, 1.0],
            [6.0, 0.0, 1.0],
            [7.0, 0.0, 1.0],
            [8.0, 0.0, 1.0],
            [9.0, 0.0, 1.0],
        ],
        'waypoints_tg': [
            [1.0, 0.0, 0.0],
            [1.0, 0.0, 0.0],
            [1.0, 0.0, 0.0],
            [1.0, 0.0, 0.0],
            [1.0, 0.0, 0.0],
            [1.0, 0.0, 0.0],
            [1.0, 0.0, 0.0],
            [1.0, 0.0, 0.0],
            [1.0, 0.0, 0.0],
            [1.0, 0.0, 0.0],
        ]
    }

    spline_generator = SplineTrajectoryGenerator(waypoints, drone_pose,3,True)
    spline_generator.first_spline()

    trajectory_points = [drone_pose['waypoints']]
    loop_count = 0
    s_eval = 0.0
    loop_count_wp = 0
    while True:
        s_eval = s_eval + spline_generator.spline_reparametrization['total_length'] * 0.1
        if spline_generator.check_status_trajectory(s_eval):
            s_eval = 0.0
            first_wp_key = list(spline_generator.waypoints.keys())[0]
            trajectory_points.append(spline_generator.waypoints[first_wp_key]['waypoints'])
            loop_count_wp += 1
        loop_count += 1
        if loop_count > 100 or loop_count_wp >= 14:  
            break
    trajectory_points = np.array(trajectory_points)

    fig = plt.figure()
    ax = fig.add_subplot(111, projection='3d')
    ax.plot(trajectory_points[:, 0], trajectory_points[:, 1], trajectory_points[:, 2], label='Trajectory')
    ax.scatter(drone_pose['waypoints'][0], drone_pose['waypoints'][1], drone_pose['waypoints'][2], color='red', label='Initial Position')
    ax.scatter( 
    [wp[0] for wp in waypoints['waypoints']],
    [wp[1] for wp in waypoints['waypoints']],
    [wp[2] for wp in waypoints['waypoints']],
        color='green', label='Waypoints'
    )
    ax.legend()
    ax.set_xlabel('X')
    ax.set_ylabel('Y')
    ax.set_zlabel('Z')
    plt.title('SplineTrajectoryGenerator')
    plt.show()

if __name__ == "__main__":
    main()

File: mpcc_trajectory/examples/spline/spline_trajectory_generator.cpp
// Copyright 2025 Universidad Politécnica de Madrid
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
//    * Redistributions of source code must retain the above copyright
//      notice, this list of conditions and the following disclaimer.
//
//    * Redistributions in binary form must reproduce the above copyright
//      notice, this list of conditions and the following disclaimer in the
//      documentation and/or other materials provided with the distribution.
//
//    * Neither the name of the Universidad Politécnica de Madrid nor the names of its
//      contributors may be used to endorse or promote products derived from
//      this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

/**
 * @file spline_trajectory_generator.cpp
 * @brief Spline trajectory generator implementation
 * @authors Rafael Pérez Seguí, Carmen De Rojas Pita-Romero
 * @copyright Copyright (c) 2025 Universidad Politécnica de Madrid
 * @license BSD-3-Clause
 */

#include "spline_trajectory_generator.hpp"
#include <cmath>

namespace trajectory
{

// ============================================================================
// Path Implementation
// ============================================================================

Path::Path(const std::vector<TrajectorySetpoint> & setpoints)
: setpoints_(setpoints), current_index_(0)
{
  if (setpoints_.empty()) {
    throw std::invalid_argument("Path requires at least one setpoint");
  }
}

std::tuple<std::vector<Eigen::Vector3d>, std::vector<Eigen::Vector3d>>
Path::getWaypoints(int num) const
{
  std::vector<Eigen::Vector3d> waypoints;
  std::vector<Eigen::Vector3d> tangents;

  for (int i = 0; i < num; ++i) {
    size_t idx = current_index_ + i;
    if (idx >= setpoints_.size()) {
      break;
    }
    waypoints.push_back(setpoints_[idx].position);
    tangents.push_back(setpoints_[idx].tangent);
  }

  return std::make_tuple(waypoints, tangents);
}

bool Path::advance()
{
  if (current_index_ + 1 < setpoints_.size()) {
    current_index_++;
    return true;
  }
  return false;
}

const TrajectorySetpoint * Path::getCurrentSetpoint() const
{
  if (current_index_ >= setpoints_.size()) {
    return nullptr;
  }
  return &setpoints_[current_index_];
}

size_t Path::remainingWaypoints() const
{
  return setpoints_.size() - current_index_;
}

// ============================================================================
// SplineTrajectoryGenerator Implementation
// ============================================================================

SplineTrajectoryGenerator::SplineTrajectoryGenerator(
  const std::vector<TrajectorySetpoint> & setpoints,
  int num_wp,
  int spline_samples,
  int spline_degree)
: num_wp_(num_wp),
  spline_samples_(spline_samples),
  spline_degree_(spline_degree),
  regeneration_count_(0),
  path_(setpoints)
{
  if (num_wp < 2) {
    throw std::invalid_argument("num_wp must be at least 2");
  }

  // Generate initial spline
  generateSpline();
}

void SplineTrajectoryGenerator::generateSpline()
{
  // Get next num_wp waypoints from Path
  auto [waypoints, tangents] = path_.getWaypoints(num_wp_);

  if (waypoints.size() < 2) {
    throw std::runtime_error("Not enough waypoints remaining to generate spline");
  }

  // Create new spline
  spline_ = std::make_unique<spline::HermiteSpline>(waypoints, tangents);

  // Compute arc length reparametrization
  spline_reparametrization_ = spline::computeArcLengthReparametrization(
    *spline_, spline_samples_, spline_degree_);

  regeneration_count_++;
}

SplineEvaluationResult SplineTrajectoryGenerator::evaluateArcLengthSpline(double s) const
{
  if (!spline_) {
    throw std::runtime_error("Spline not initialized");
  }

  // Normalize s to [0, 1]
  double s_norm = std::clamp(
    s / spline_reparametrization_.total_length, 0.0, 1.0);

  // Polynomial approximation: t(s_norm)
  double t = 0.0;
  int n_coeffs = spline_reparametrization_.poly_coeffs.size();
  for (int i = 0; i < n_coeffs; ++i) {
    t += spline_reparametrization_.poly_coeffs[i] *
      std::pow(s_norm, n_coeffs - 1 - i);
  }

  // Evaluate spline at parameter t
  Eigen::Vector3d position = spline_->evaluate(t);
  Eigen::Vector3d derivative = spline_->evaluateDerivative(t);

  SplineEvaluationResult result;
  result.position = position;
  result.derivative = derivative;
  result.theta = s_norm;
  result.reparametrization = spline_reparametrization_;

  return result;
}

std::tuple<double, SplineEvaluationResult>
SplineTrajectoryGenerator::evaluateSpline(double s)
{
  // Evaluate spline at current arc length s
  auto result = evaluateArcLengthSpline(s);

  double new_s = s;

  // Check if we've passed the end of current spline
  if (result.theta >= 1.0) {
    // Advance to next waypoint
    if (path_.advance()) {
      // Regenerate spline with new waypoints
      try {
        generateSpline();
        // Reset s to 0 for the new spline
        new_s = 0.0;
        // Re-evaluate at s=0
        result = evaluateArcLengthSpline(new_s);
      } catch (const std::runtime_error & e) {
        // Not enough waypoints remaining, keep current spline
        // Keep s as is (will be clamped to spline length)
      }
    }
    // else: No more waypoints available, keep current spline
  }

  return std::make_tuple(new_s, result);
}

}  // namespace trajectory
